client/src/app/editor/file-outputs/file-outputs.component.ts

```diff
@@ -57,30 +57,79 @@ export class FileOutputsComponent implements OnChanges, OnInit {
   ngOnInit() {
     const outputFileId = this.route.snapshot.queryParamMap.get('preview');
 
-    if (outputFileId) {
-      this.hasOutput
-        .pipe(
-          filter(hasOutput => hasOutput),
-          mergeMap(() => this.dataSource.connect()),
-          mergeMap(outputFiles => outputFiles),
-          filter(outputFile => outputFile.id === outputFileId),
-          filter(outputFile => isImage(outputFile.name)),
-          take(1)
+    const executionSelected$ = this.executionId$.pipe(
+      switchMap(executionId => this.labExecutionService.executionExists(executionId).pipe(catchError(() => of(false)))),
+      filter(Boolean)
+    );
+
+    const executionsAdded$ = this.labExecutionService.observeExecutionsForLab(this.lab).pipe(
+      mergeMap(executions => executions),
+      filter(labExecution => labExecution.id === this.executionId),
+      distinctUntilChanged((a, b) => a.id === b.id),
+      // We skip the first value because on page load 'executionSelected$' already emits the
+      // currently selected execution we want to observe. 'executionsAdded$' should only emit
+      // an execution if the user has started a new one. See comment below on the behavior of
+      // 'observeExecution'.
+      skip(1)
+    );
+
+    // We have to merge those two streams together because somehow, when 'observeExecution' is called
+    // on an execution that was just started, it will never emit a value, although it uses ref.value()
```

Comment:
So this comment sounds more like "somehow it's doing something we don't understand, so we just use this hack".

But wasn't it because we were listening on executions with id that simply didn't exist at the time of listening yet, which is why it never emits, even if that execution with that id will be added to the database at some point?